bootenv/keydata_test.go

```diff
@@ -0,0 +1,457 @@
+	return testutil.MakeMockCore20ModelAssertion(c, map[string]interface{}{
+		"authority-id": "fake-brand",
+		"series":       "16",
+		"brand-id":     "fake-brand",
+		"model":        modelName,
+		"grade":        "secured",
+	}, "Jv8_JiHiIzJVcO9M55pPdqSDWUvuhfDIBJUS-3VW7F_idjix7Ffn5qMxB21ZQuij")
+}
+
+func (s *keyDataPlatformSuite) TestBootEnvAuthStateErrors(c *C) {
+	primaryKey := s.newPrimaryKey(c, 32)
+
+	params := &KeyDataScopeParams{
+		PrimaryKey: primaryKey,
+		Role:       "test",
+		KDFAlg:     crypto.SHA256,
+		MDAlg:      crypto.SHA256,
+		ModelAlg:   crypto.SHA256,
+	}
+
+	kds, err := NewKeyDataScope(params)
+	c.Assert(err, IsNil)
+
+	authModels := []SnapModel{
+		s.makeMockModelAssertion(c, "fake-model"),
+	}
+
+	c.Check(kds.SetAuthorizedSnapModels(primaryKey, params.Role, authModels...), IsNil)
+
+	err = kds.IsBootEnvironmentAuthorized()
+	c.Check(err, ErrorMatches, "SetModel hasn't been called yet")
+
+	SetModel(authModels[0])
+
+	authModes := []string{
+		"modeFoo",
+	}
+
+	c.Check(kds.SetAuthorizedBootModes(primaryKey, params.Role, authModes...), IsNil)
+	err = kds.IsBootEnvironmentAuthorized()
+	c.Check(err, ErrorMatches, "SetBootMode hasn't been called yet")
+}
+
+type testSetAuthorizedSnapModelsData struct {
+	kDFAlg      crypto.Hash
+	mDAlg       crypto.Hash
+	modelAlg    crypto.Hash
+	validRole   string
+	role        string
+	validModels []SnapModel
+}
+
+func (s *keyDataPlatformSuite) testSetAuthorizedSnapModels(c *C, data *testSetAuthorizedSnapModelsData) error {
+	primaryKey := s.newPrimaryKey(c, 32)
+
+	params := &KeyDataScopeParams{
+		PrimaryKey: primaryKey,
+		Role:       data.validRole,
+		KDFAlg:     data.kDFAlg,
+		MDAlg:      data.mDAlg,
+		ModelAlg:   data.modelAlg,
+	}
+
+	kds, err := NewKeyDataScope(params)
+	c.Assert(err, IsNil)
+
+	err = kds.SetAuthorizedSnapModels(primaryKey, data.role, data.validModels...)
+	return err
+}
+
+func (s *keyDataPlatformSuite) TestSetAuthorizedSnapModels(c *C) {
+	validModels := []SnapModel{
+		s.makeMockModelAssertion(c, "model-a"),
+	}
+	c.Check(
+		s.testSetAuthorizedSnapModels(c, &testSetAuthorizedSnapModelsData{
+			kDFAlg:      crypto.SHA256,
+			mDAlg:       crypto.SHA256,
+			modelAlg:    crypto.SHA256,
+			validRole:   "test",
+			role:        "test",
+			validModels: validModels,
+		}), IsNil)
+}
+
+func (s *keyDataPlatformSuite) TestSetAuthorizedSnapModelsInvalidRole(c *C) {
+	// test authorization error when SetAuthorizedSnapModels is called with
+	// a role different than the one set in its keyDataScope.
+	validModels := []SnapModel{
+		s.makeMockModelAssertion(c, "model-a"),
+	}
+	c.Check(
+		s.testSetAuthorizedSnapModels(c, &testSetAuthorizedSnapModelsData{
+			kDFAlg:      crypto.SHA256,
+			mDAlg:       crypto.SHA256,
+			modelAlg:    crypto.SHA256,
+			validRole:   "test",
+			role:        "different",
+			validModels: validModels,
+		}), ErrorMatches, "incorrect key supplied")
+}
```

Comment:
There doesn't appear to be a test of what happens if the wrong key is supplied to `SetAuthorizedSnapModels`.

Reply:
Added.